### ISSUE TITLE
Many small metadata writes causes bottleneck in DB

### ISSUE THREAD

**Issue by geoffjentry** [broadinstitute/cromwell#1810]
While benchmarking some performance enhancements I've been playing with I kept noticing that no matter how fast I could get things eventually performance would drop back down to baseline levels from develop. I traced it down to the `WriteMetadataActor`, specifically it appears that writing boatloads of individual events (not atypical under load) can cause a lot of problems (not particularly surprising, but ...)

It seems like some sort of batching/work pulling scheme could work wonders here although presumably it'd then come at the cost of memory (to buffer the unwritten values). That's just one thought, not a prescription

**Comment by kcibul** [comment]
Are we using single inserts or batch inserts when writing to the DB?  That can cause huge performance gains due the reduction in chatter

**Comment by mcovarr** [comment]
We are using single inserts.  We should be using batch inserts like Rawls per placeholder ticket #1582.

**Comment by geoffjentry** [comment]
Is that a db level setting giving you bulk insert via magic? 

Doing it by hand turns out to not be super simple to make it not suck when it comes to our metadata, although I think I came up with something that would work after I quit poking at it. I won't bother fiddling with it if there's potential magic though

Btw I think this might be behind a lot of the various things I've been tracking, the pattern is pretty much the same. Well, more generally I think this is a pattern we use a lot (not just the single writes but firing these things off as fast as possible due to the futures) but this is the #1 culprit by far

**Comment by geoffjentry** [comment]
Btw it isn't *entirely* true that it's only using single writes but it might as well be considered that way

**Comment by geoffjentry** [comment]
To put some concrete numbers on this, I'm using `ab` to submit workflows using 32 threads. 

Using `develop` I get a sustained ~450 requests per second for as long as my patience lasts (I've not gone beyond 200k submissions). CPU utilization remains steady throughout.

Using my changes for small bursts I get ~2500 rps, if I run 200k submits I get ~1400 rps in aggregate. It starts going down pretty rapidly after a few minutes and not long after that is down to roughly the rate of develop (if that) - by 400k submissions it gets too slow for me to wait around. You can see the impact in the CPU utilization of both the JVM and MySQL as well.

Using my changes and removing the metadata write I get a sustained 4200 requests per second up until at least 800k workflows. CPU utilization remains steady throughout.

**Comment by geoffjentry** [comment]
Gave up trying to combat the performance vs memory tradeoff (which seems unlikely to be winnable anyways) and the relatively simple solution works. We can have the metadata write batch size be a config option and tell users that setting the number lower uses less memory but drops performance under higher load and vice versa

**Comment by kcibul** [comment]
Cool -- so this is JDBC-batching effectively?  It might be good to turn on SQL logging and check to see that the number of writes match what you think they should be.  I remember @dvoet saying that the query rewrites weren't happening when they thought they should and that there was an option they had to set (either CloudSQL or Slick, can't remember).  We should check in with him after break

**Comment by geoffjentry** [comment]
@kcibul I don't know. Because the ticket @mcovarr linked sounded like there was a magic setting somewhere I googled around a bit and found some references to badness. I didn't check, however. Still something to look at. Also I've been using MySQL not CloudSQL, perhaps that matters.

I got to the point where if I set my batch size high enough (I was generating on the order of ~15k events to write per second, FWIW) my overall performance was such that I was getting a sustained rate of ~1500-1700 (I forget exactly) requests per second on the submission side, which is certainly still a lot less than I was getting w/o metadata at all but a heck of a lot better than I was able to do otherwise.
